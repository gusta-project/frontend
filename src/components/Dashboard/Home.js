import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import React, { Component, Fragment } from 'react';
import { bindActionCreators } from 'redux';
import { Helmet } from 'react-helmet';
import { CardColumns, Card } from 'react-bootstrap';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';

import { actions as dashboardActions } from 'reducers/dashboard';
import { getStats } from 'selectors/dashboard';

export class DashboardHome extends Component {
  static propTypes = {
    actions: PropTypes.object.isRequired,
    stats: PropTypes.object.isRequired
  };

  constructor(props) {
    super(props);
    this.refresh = this.refresh.bind(this);
  }

  refresh() {
    const { actions } = this.props;

    actions.requestStats();
  }

  render() {
    const { stats } = this.props;
    const {
      activatedUsers,
      flavors,
      flavorTags,
      recipes,
      recipeTags,
      tags,
      users,
      userTokens,
      vendors
    } = stats;

    return (
      <Fragment>
        <CardColumns>
          <Helmet title="Dashboard" />
          <Card bg="danger" text="white">
            <Card.Header>Users: {users}</Card.Header>
            <Card.Body>
              <h3>Activated Users: {activatedUsers}</h3>
              <h3>Inactive Users: {users - activatedUsers}</h3>
              <h3>Active Tokens: {userTokens}</h3>
            </Card.Body>
          </Card>
          <Card bg="primary" text="white">
            <Card.Body>
              <h2>Tags: {tags}</h2>
            </Card.Body>
          </Card>
          <Card bg="secondary" text="white">
            <Card.Body>
              <h2>Vendors: {vendors}</h2>
            </Card.Body>
          </Card>
          <Card bg="info" text="white">
            <Card.Header>Flavors: {flavors}</Card.Header>
            <Card.Body>
              <h2>Tags: {flavorTags}</h2>
            </Card.Body>
          </Card>
          <Card bg="warning" text="white">
            <Card.Header>Recipes: {recipes}</Card.Header>
            <Card.Body>
              <h2>Tags: {recipeTags}</h2>
            </Card.Body>
          </Card>
        </CardColumns>
        <a href="#home" onClick={this.refresh} title="Refresh Stats">
          <FontAwesomeIcon icon="sync" />
        </a>
      </Fragment>
    );
  }
}

const mapStateToProps = state => ({
  stats: getStats(state)
});

const mapDispatchToProps = dispatch => ({
  actions: bindActionCreators(dashboardActions, dispatch)
});

export default connect(mapStateToProps, mapDispatchToProps)(DashboardHome);
