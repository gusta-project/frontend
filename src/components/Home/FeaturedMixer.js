import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Row, Col, Card, Button } from 'react-bootstrap';

export class FeaturedMixer extends Component {
  render() {
    return (
      <Row className="text-center">
        <Col>
          <h2>Featured mixer of the month</h2>
          <Card className="text-center">
            <Card.Header>mlNikon</Card.Header>
            <Card.Body>
              <Row>
                <Col>
                  <img src="/media/harold.jpeg" alt="" className="w-50" />
                </Col>
                <Col>
                  <Card.Title>I&apos;m not creative enough</Card.Title>
                  <Card.Text>
                    To figure out what could go here, but it&apos;s an idea.
                  </Card.Text>
                  <Button variant="primary" className="button-animation">
                    <span>Check out their recipes</span>
                  </Button>
                </Col>
              </Row>
            </Card.Body>
            <Card.Footer className="text-muted">
              <span>The next featured mixer will be here in 5 days</span>
            </Card.Footer>
          </Card>
        </Col>
      </Row>
    );
  }
}

export default connect(null, null)(FeaturedMixer);
