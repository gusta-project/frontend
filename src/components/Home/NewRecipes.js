import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import { Row, Col, CardGroup, Card } from 'react-bootstrap';

export class NewRecipes extends Component {
  render() {
    return (
      <Row className="text-center">
        <Col>
          <h2>Newest recipes</h2>
          <CardGroup>
            <Card>
              <Card.Img
                variant="top"
                src="/media/card-test-1.jpg"
                className="img-fluid w-75 mx-auto"
              />
              <Card.Body>
                <Card.Title>
                  <Link to="/user/david">Strawberry Milkshake</Link>
                </Card.Title>
                <Card.Text>by David Dyess</Card.Text>
              </Card.Body>
              <Card.Footer>
                <small>
                  <span role="img" aria-label="Checkmark emoji">
                    ✔
                  </span>
                  ️ You have everything!
                </small>
              </Card.Footer>
            </Card>
            <Card>
              <Card.Img
                variant="top"
                src="/media/card-test-2.jpg"
                className="img-fluid w-75 mx-auto"
              />
              <Card.Body>
                <Card.Title>
                  <Link to="#">Bomb Pop</Link>
                </Card.Title>
                <Card.Text>by JosefBud</Card.Text>
              </Card.Body>
              <Card.Footer>
                <small>
                  <span role="img" aria-label="Checkmark emoji">
                    ✔
                  </span>
                  ️ You have everything!
                </small>
              </Card.Footer>
            </Card>
            <Card>
              <Card.Img
                variant="top"
                src="/media/card-test-3.jpg"
                className="img-fluid w-75 mx-auto"
              />
              <Card.Body>
                <Card.Title>
                  <Link to="#">Tiramisu</Link>
                </Card.Title>
                <Card.Text>by ayan4m1</Card.Text>
              </Card.Body>
              <Card.Footer>
                <small>
                  <span role="img" aria-label="X emoji">
                    ❌
                  </span>{' '}
                  You are missing 1 flavor
                </small>
              </Card.Footer>
            </Card>
            <Card>
              <Card.Img
                variant="top"
                src="/media/card-test-4.jpeg"
                className="img-fluid w-75 mx-auto"
              />
              <Card.Body>
                <Card.Title>
                  <Link to="#">RY4 Tobaccy</Link>
                </Card.Title>
                <Card.Text>by pscn</Card.Text>
              </Card.Body>
              <Card.Footer>
                <small>
                  <span role="img" aria-label="Checkmark emoji">
                    ✔
                  </span>
                  ️ You have everything!
                </small>
              </Card.Footer>
            </Card>
            <Card>
              <Card.Img
                variant="top"
                src="/media/card-test-5.jpeg"
                className="img-fluid w-75 mx-auto"
              />
              <Card.Body>
                <Card.Title>
                  <Link to="#">Fruit Punch</Link>
                </Card.Title>
                <Card.Text>by Korlimann</Card.Text>
              </Card.Body>
              <Card.Footer>
                <small>
                  <span role="img" aria-label="X emoji">
                    ❌
                  </span>{' '}
                  You are missing 3 flavors
                </small>
              </Card.Footer>
            </Card>
          </CardGroup>
        </Col>
      </Row>
    );
  }
}

export default connect(null, null)(NewRecipes);
