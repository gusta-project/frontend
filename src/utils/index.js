import dayjs from 'dayjs';

import { getApplication, getAuthorization } from 'selectors/application';

/* eslint-disable react/display-name, react/prop-types */

/**
 * Creates an object containing action constants namespaced under the specified reducer.
 *
 * @param {string} reducer The name of the reducer or module
 * @param {string[]} actions A list of action names
 * @return {object} Object with action name keys and full action string values
 */
export function buildActions(reducer, actions) {
  const result = {};

  for (const action of actions) {
    result[action] = `${reducer}/${action}`;
  }

  return result;
}

/**
 * Interpolate strings from the `params` object into the `url`.
 *
 * @param {object} endpoint Object with `url` and optional `params` object
 */
export const buildUrl = (endpoint) => {
  // this is defined by webpack.DefinePlugin
  const baseUrl = API_URL;
  const { url, params } = endpoint;

  if (!params) {
    return `${baseUrl}${url}`;
  }

  return Object.entries(params).reduce((resultUrl, entry) => {
    const [key, value] = entry;

    return resultUrl.replace(`{${key}}`, value);
  }, `${baseUrl}${url}`);
};

/**
 * Gets the initial Redux state from localStorage and the source code defaults.
 */
export const getInitialState = () => {
  // this needs to be require()d because an import results in a
  // circular dependency
  const { initialState } = require('reducers');

  try {
    const [accessToken, expiration] = [
      localStorage.getItem('accessToken'),
      localStorage.getItem('expiration')
    ];

    if (!accessToken) {
      return initialState;
    }

    const expirationDate = dayjs(JSON.parse(expiration));

    if (!expirationDate.isValid() || expirationDate.isBefore(dayjs())) {
      localStorage.removeItem('accessToken');
      localStorage.removeItem('expiration');
      return initialState;
    }

    return {
      ...initialState,
      application: {
        ...getApplication(initialState),
        authorization: {
          ...getAuthorization(initialState),
          accessToken: JSON.parse(accessToken),
          expiration: expirationDate
        }
      }
    };
  } catch {
    return initialState;
  }
};
