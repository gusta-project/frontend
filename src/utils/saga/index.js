import counterUtilities from './counter';
import pageUtilities from './pager';
import toastUtilities from './toasts';

export default {
  ...counterUtilities,
  ...pageUtilities,
  ...toastUtilities
};
