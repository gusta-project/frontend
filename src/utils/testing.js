import React, { createElement } from 'react';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';

/**
 * Creates a mock component which takes a set of props specified at runtime.
 *
 * @param {string} name The component/element name (in PascalCase)
 * @param {object} props An object containing props to setup
 * @return {object} Mock component with specified name and props
 */
export const mockComponent = (name, props = {}) => () =>
  createElement(name, props, props.children);

/**
 * Creates a mock component which will expose its props for snapshot testing purposes
 * @param {*} name The component/element name (in PascalCase)
 */
export const mockComponentWithProps = (name) => (props) =>
  createElement(name, props, props.children);

/**
 * Wraps a React component in a <MemoryRouter> suitable for testing.
 *
 * @param {Component} WrappedComponent React component to wrap
 */
export const withMemoryRouter = (WrappedComponent, routerProps = {}) => (
  props
) => (
  <MemoryRouter {...routerProps}>
    <WrappedComponent {...props} />
  </MemoryRouter>
);

/**
 * Wraps a React component in a <Provider> suitable for testing.
 *
 * @param {Component} WrappedComponent React component to wrap
 * @param {object} store Mock store
 */
export const withProvider = (WrappedComponent, store) => (props) => (
  <Provider store={store}>
    <WrappedComponent {...props} />
  </Provider>
);
