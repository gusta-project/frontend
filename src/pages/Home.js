import { Helmet } from 'react-helmet';
import React, { Component } from 'react';
import { Container, Row, Col } from 'react-bootstrap';
import { Link } from 'react-router-dom';

import TopRecipesDay from 'components/Home/TopRecipesDay';
import NewRecipes from 'components/Home/NewRecipes';
import FeaturedMixer from 'components/Home/FeaturedMixer';

export default class Home extends Component {
  render() {
    return (
      <Container>
        <Helmet title="Home" />
        <Row className="text-center">
          <Col>
            <h1>Welcome to MixNJuice!</h1>
            <p style={{ fontSize: '0.8em' }}>
              <Link to="#">Click here to customize your front page!</Link>
            </p>
          </Col>
        </Row>
        <TopRecipesDay />
        <hr />
        <NewRecipes />
        <hr />
        <FeaturedMixer />
      </Container>
    );
  }
}
