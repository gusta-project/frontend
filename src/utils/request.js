import axios from 'axios';
import { buildUrl } from 'utils';
import { call, race, delay } from 'redux-saga/effects';

/**
 * Make an XMLHttpRequest to the specified endpoint with the specified request data.
 *
 * @param {object} endpoint An object with `url` and `method` keys
 * @param {object} data An object to use as the request body
 */
class Request {
  *execute({ endpoint, data, options = {} }) {
    try {
      // ensure the endpoint was supplied
      if (!endpoint) {
        throw new Error('No endpoint provided!');
      }

      // ensure properties required by axios are present
      const { url, method } = endpoint;

      if (!url) {
        throw new Error('Endpoint is missing URL!');
      }

      if (!method) {
        throw new Error('Endpoint is missing method!');
      }

      // default timeout of 30 seconds
      const { timeout = 30000 } = options;
      const requestUrl = yield call(buildUrl, endpoint);

      // start a race between the request and a timer, cancel the loser
      const { response } = yield race({
        response: call(axios, {
          url: requestUrl,
          method,
          data
        }),
        timeout: delay(timeout)
      });

      if (!response) {
        throw new Error('Request timed out!');
      }

      const { status } = response;
      const success = status >= 200 && status < 400;

      if (success) {
        return {
          success,
          response
        };
      } else {
        return {
          success,
          error: response
        };
      }
    } catch (error) {
      return {
        success: false,
        error
      };
    }
  }
}

const instance = new Request();

export default instance;
