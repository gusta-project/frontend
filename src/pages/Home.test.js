import { render } from '@testing-library/react';
import React from 'react';
import configureStore from 'redux-mock-store';

import Home from './Home';
import { withMemoryRouter, withProvider } from 'utils/testing';

describe('<Home />', () => {
  const initialState = {};
  const mockStore = configureStore();
  const store = mockStore(initialState);

  it('renders correctly', () => {
    const RoutedHome = withMemoryRouter(Home);
    const ConnectedHome = withProvider(RoutedHome, store);
    const { getByText } = render(<ConnectedHome />);

    expect(getByText('Welcome to MixNJuice!')).toBeInTheDocument();
  });
});
