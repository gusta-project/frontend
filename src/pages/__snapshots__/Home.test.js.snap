// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Home /> renders correctly 1`] = `
<div
  className="container"
>
  <div
    className="text-center row"
  >
    <div
      className="col"
    >
      <h1>
        Welcome to MixNJuice!
      </h1>
      <p
        style={
          Object {
            "fontSize": "0.8em",
          }
        }
      >
        <a
          href="/"
          onClick={[Function]}
        >
          Click here to customize your front page!
        </a>
      </p>
    </div>
  </div>
  <div
    className="text-center row"
  >
    <div
      className="col"
    >
      <h2>
        Top recipes of the day
      </h2>
      <div
        className="card-group"
      >
        <div
          className="card"
        >
          <img
            className="card-img-top img-fluid w-75 mx-auto"
            src="/media/card-test-1.jpg"
          />
          <div
            className="card-body"
          >
            <div
              className="card-title h5"
            >
              <a
                href="/"
                onClick={[Function]}
              >
                Strawberry Milkshake
              </a>
            </div>
            <p
              className="card-text"
            >
              by David Dyess
            </p>
          </div>
          <div
            className="card-footer"
          >
            <small>
              <span
                aria-label="Checkmark emoji"
                role="img"
              >
                ✔
              </span>
              ️ You have everything!
            </small>
          </div>
        </div>
        <div
          className="card"
        >
          <img
            className="card-img-top img-fluid w-75 mx-auto"
            src="/media/card-test-2.jpg"
          />
          <div
            className="card-body"
          >
            <div
              className="card-title h5"
            >
              <a
                href="/"
                onClick={[Function]}
              >
                Bomb Pop
              </a>
            </div>
            <p
              className="card-text"
            >
              by JosefBud
            </p>
          </div>
          <div
            className="card-footer"
          >
            <small>
              <span
                aria-label="Checkmark emoji"
                role="img"
              >
                ✔
              </span>
              ️ You have everything!
            </small>
          </div>
        </div>
        <div
          className="card"
        >
          <img
            className="card-img-top img-fluid w-75 mx-auto"
            src="/media/card-test-3.jpg"
          />
          <div
            className="card-body"
          >
            <div
              className="card-title h5"
            >
              <a
                href="/"
                onClick={[Function]}
              >
                Tiramisu
              </a>
            </div>
            <p
              className="card-text"
            >
              by ayan4m1
            </p>
          </div>
          <div
            className="card-footer"
          >
            <small>
              <span
                aria-label="X emoji"
                role="img"
              >
                ❌
              </span>
               
              You are missing 1 flavor
            </small>
          </div>
        </div>
        <div
          className="card"
        >
          <img
            className="card-img-top img-fluid w-75 mx-auto"
            src="/media/card-test-4.jpeg"
          />
          <div
            className="card-body"
          >
            <div
              className="card-title h5"
            >
              <a
                href="/"
                onClick={[Function]}
              >
                RY4 Tobaccy
              </a>
            </div>
            <p
              className="card-text"
            >
              by pscn
            </p>
          </div>
          <div
            className="card-footer"
          >
            <small>
              <span
                aria-label="Checkmark emoji"
                role="img"
              >
                ✔
              </span>
              ️ You have everything!
            </small>
          </div>
        </div>
        <div
          className="card"
        >
          <img
            className="card-img-top img-fluid w-75 mx-auto"
            src="/media/card-test-5.jpeg"
          />
          <div
            className="card-body"
          >
            <div
              className="card-title h5"
            >
              <a
                href="/"
                onClick={[Function]}
              >
                Fruit Punch
              </a>
            </div>
            <p
              className="card-text"
            >
              by Korlimann
            </p>
          </div>
          <div
            className="card-footer"
          >
            <small>
              <span
                aria-label="X emoji"
                role="img"
              >
                ❌
              </span>
               
              You are missing 3 flavors
            </small>
          </div>
        </div>
      </div>
    </div>
  </div>
  <hr />
  <div
    className="text-center row"
  >
    <div
      className="col"
    >
      <h2>
        Newest recipes
      </h2>
      <div
        className="card-group"
      >
        <div
          className="card"
        >
          <img
            className="card-img-top img-fluid w-75 mx-auto"
            src="/media/card-test-1.jpg"
          />
          <div
            className="card-body"
          >
            <div
              className="card-title h5"
            >
              <a
                href="/user/david"
                onClick={[Function]}
              >
                Strawberry Milkshake
              </a>
            </div>
            <p
              className="card-text"
            >
              by David Dyess
            </p>
          </div>
          <div
            className="card-footer"
          >
            <small>
              <span
                aria-label="Checkmark emoji"
                role="img"
              >
                ✔
              </span>
              ️ You have everything!
            </small>
          </div>
        </div>
        <div
          className="card"
        >
          <img
            className="card-img-top img-fluid w-75 mx-auto"
            src="/media/card-test-2.jpg"
          />
          <div
            className="card-body"
          >
            <div
              className="card-title h5"
            >
              <a
                href="/"
                onClick={[Function]}
              >
                Bomb Pop
              </a>
            </div>
            <p
              className="card-text"
            >
              by JosefBud
            </p>
          </div>
          <div
            className="card-footer"
          >
            <small>
              <span
                aria-label="Checkmark emoji"
                role="img"
              >
                ✔
              </span>
              ️ You have everything!
            </small>
          </div>
        </div>
        <div
          className="card"
        >
          <img
            className="card-img-top img-fluid w-75 mx-auto"
            src="/media/card-test-3.jpg"
          />
          <div
            className="card-body"
          >
            <div
              className="card-title h5"
            >
              <a
                href="/"
                onClick={[Function]}
              >
                Tiramisu
              </a>
            </div>
            <p
              className="card-text"
            >
              by ayan4m1
            </p>
          </div>
          <div
            className="card-footer"
          >
            <small>
              <span
                aria-label="X emoji"
                role="img"
              >
                ❌
              </span>
               
              You are missing 1 flavor
            </small>
          </div>
        </div>
        <div
          className="card"
        >
          <img
            className="card-img-top img-fluid w-75 mx-auto"
            src="/media/card-test-4.jpeg"
          />
          <div
            className="card-body"
          >
            <div
              className="card-title h5"
            >
              <a
                href="/"
                onClick={[Function]}
              >
                RY4 Tobaccy
              </a>
            </div>
            <p
              className="card-text"
            >
              by pscn
            </p>
          </div>
          <div
            className="card-footer"
          >
            <small>
              <span
                aria-label="Checkmark emoji"
                role="img"
              >
                ✔
              </span>
              ️ You have everything!
            </small>
          </div>
        </div>
        <div
          className="card"
        >
          <img
            className="card-img-top img-fluid w-75 mx-auto"
            src="/media/card-test-5.jpeg"
          />
          <div
            className="card-body"
          >
            <div
              className="card-title h5"
            >
              <a
                href="/"
                onClick={[Function]}
              >
                Fruit Punch
              </a>
            </div>
            <p
              className="card-text"
            >
              by Korlimann
            </p>
          </div>
          <div
            className="card-footer"
          >
            <small>
              <span
                aria-label="X emoji"
                role="img"
              >
                ❌
              </span>
               
              You are missing 3 flavors
            </small>
          </div>
        </div>
      </div>
    </div>
  </div>
  <hr />
  <div
    className="text-center row"
  >
    <div
      className="col"
    >
      <h2>
        Featured mixer of the month
      </h2>
      <div
        className="text-center card"
      >
        <div
          className="card-header"
        >
          mlNikon
        </div>
        <div
          className="card-body"
        >
          <div
            className="row"
          >
            <div
              className="col"
            >
              <img
                alt=""
                className="w-50"
                src="/media/harold.jpeg"
              />
            </div>
            <div
              className="col"
            >
              <div
                className="card-title h5"
              >
                I'm not creative enough
              </div>
              <p
                className="card-text"
              >
                To figure out what could go here, but it's an idea.
              </p>
              <button
                className="button-animation btn btn-primary"
                disabled={false}
                type="button"
              >
                <span>
                  Check out their recipes
                </span>
              </button>
            </div>
          </div>
        </div>
        <div
          className="text-muted card-footer"
        >
          <span>
            The next featured mixer will be here in 5 days
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;
